[Single-Exe] Run from bundle

swaroop-sridhar opened the pull request:
## Changes

corebundle Host:
 - Remove code that deals with extraction to temporary dirs.
 - Process the bundle manifest and export a callback to look for files in the bundle.
 - The TPA doesn't contain bundled assemblies any more.

CoreCLR Initialization:
 - Setup BundleInfo structure, and pass it to appropriate initializers
 - Expect to find System.Private.Corelib in the bundle

Bundle Info:
  - Holds information about the bundle
  - Exports a wrapper around the bundle runner's probe function for path adjustments
  - Performs Unicode/UTF8 translations as necessary (because host uses UTF8 on Unix)

AppDomain:
  - Add methods obtain BundleInfo

Binder:
  - Look for assemblies in the Bundle before searching the TPA

PEImage:
  - FlatImageLayout and MappedImageLayout can take files embedded within the bundle.
  - Memory mapping routines can take files embedded within the bundle.

A few other fixes necessary to build/run.

## Testing

Tested templates: console, webapi

## Performance

### Console Hello World

No noticable change in execution time.

Run from (normal) publish directory: 104ms
Run from single-file: 99ms

### Startup of WebApiTemplate:
Measured using https://github.com/rynowak/link-a-thon on a local (real) machine.

Both cases had runs with similar timing; but the average startup was consistently slightly faster for single-exe runs.

Run from (normal) publish directory:
801.8271
782.9489
790.8015
798.7449
797.3141
783.0968
790.0275
786.7621
802.2383
778.9866
Average startup time (ms): 791.27478

Run from single-file:
742.1464
692.1949
795.1612
798.4845
809.4075
794.2644
790.6254
634.1305
686.2718
806.5665
Average startup time (ms): 754.92531

## Steps to build single-exe `app`

The following steps will work for most apps, minor adjustments may be necessary.
* `dotnet publish -r linux-x64 -c release -o pub`
* `cp pub/*.dll source/`
* `cp coreclr/bin/Product/Linux.x64.Release/corebundle source/<app>`
* `cp coreclr/bin/Product/Linux.x64.Release/System.Private.Corelib.dll source/`
* `dotnet coreclr/bin/Product/Linux.x64.Release/bundle.dll --source source --host <app>`
* Run `app`

vitek-karas commented on src/binder/assemblybinder.cpp:
```diff
@@ -1158,13 +1166,48 @@ namespace BINDER_SPACE
+            // Is assembly in the bundle?
+            // Single-file bundle contents take precedence over TPA.
+            // Bundled assemblies are treated similar to TPAs.
+            if (g_BinderVariables->fIsBundle)
```
I'm wondering why do we even need this. The TPA should contain the bundle assemblies anyway - and the code below to handle the TPA will end up calling the GetAssembly with the TPA file path. Currently the BundleInfo will actually deal with full file paths as well (it matches the prefix), but even if it would not, it would make sense to have the TPA modified to contain the "bundle paths".

Since you implemented the bundle hookup in the PEImage all the callers to the PEImage can be pretty much oblivious to whether the file comes from bundle or not.

swaroop-sridhar replied:
The bundle meta-data is not reflected in the TPA. For self-contained single-file apps, typically the TPA is just the `app.dll`. 

I don't think we should add the bundled contents to the TPA -- it is unnecessary to search for the bundle contents in two places. In a full implementation, it is better to retire the TPA and use a common host-callback instead.

vitek-karas replied:
Thanks I think I finally understand the change 😄 
Can you please extend the comment to mention:
* TPA is intentionally incomplete for bundles
* Bundled assemblies are not listed in TPA - and so we first probe the bundle
* The code will call `GetAssembly` with just a file name, no path

janvorli commented on src/coreclr/hosts/CMakeLists.txt:
```diff
@@ -1,4 +1,5 @@
 include_directories(inc)
+include_directories(../../inc)
```
I was wondering why we need this added include and so I've tried to compile without this change - and it was ok both on Linux and Windows.

jkotas commented on src/inc/MSCOREE.IDL:
```diff
@@ -221,7 +224,7 @@ typedef enum
 interface ICLRRuntimeHost : IUnknown
 {
     // Starts the runtime. This is equivalent to CoInitializeCor().
```
This is legacy hosting interface that number of folks out there still depends on. It is even mentioned in the docs: https://docs.microsoft.com/en-us/dotnet/core/tutorials/netcore-hosting

 Avoid changing it

jkotas commented on src/vm/ceemain.h:
```diff
@@ -16,14 +16,15 @@
 
 // Ensure the EE is started up.
-HRESULT EnsureEEStarted(COINITIEE flags);
+HRESULT EnsureEEStarted(COINITIEE flags, const BundleInfo* bundleInfo);
```
Do you really need to pass the BundleInfo* everywhere? Can we just store it in a static and get everybody who needs to fetch it from the static?

vitek-karas replied:
:+1 at the very least we should pass it along at the same place where we pass runtime properties - which is `CreateAppDomainWithManager`, but ideally this would not go through the COM APIs at all. As mentioned in other comments, I think this should be stored only in one place - the AppDomain is probably as good a place as any right now.

swaroop-sridhar commented:
>> Do you have a sample how to create a bundle like this which can be tested?

I've added the steps to the PR description

swaroop-sridhar commented on src/md/CMakeLists.txt:
```diff
@@ -1,3 +1,7 @@
+if(NOT DEFINED ENV{ROOTFS_DIR})
+  include_directories(SYSTEM /usr/local/include)
+endif()
```
Sorry this is residual from an earlier version my change; I've removed it now.

swaroop-sridhar commented on src/dlls/mscoree/unixinterface.cpp:
```diff
@@ -168,6 +169,7 @@ int coreclr_initialize(
             int propertyCount,
             const char** propertyKeys,
             const char** propertyValues,
+            bool bundleProbe(const char*, int64_t*, int64_t*),
```
We decided to not to try-design the Host-Runtime interface as part of the prototype. However, I do want to have the discussion about how the interface will look in the final scheme.

I do want to clean up the APIs at least to the point of test-pass. I'll start with the change to not pass around BundleInfo as @jkotas suggested.

swaroop-sridhar commented on src/vm/corbundle.cpp:
```diff
@@ -0,0 +1,80 @@
+    //  BundleInfo.Probe("path/to/exe/lib.dll") => m_probe("lib.dll")
+    //  BundleInfo.Probe("path/to/exe/and/some/more/lib.dll") => m_probe("and/some/more/lib.dll")
+    //
+    // This strategy obviously hides any actual files in path/to/exe from being loaded.
```
For the prototype, I think we should not have the same assembly in the two locations: bundle and next to the single-file. 

If we want to process some assemblies from actual files (for debugging), it is better to exclude them from the bundle (especially since bundle construction is manual in the prototype).

swaroop-sridhar commented on src/vm/peimagelayout.cpp:
```diff
@@ -599,12 +606,21 @@ FlatImageLayout::FlatImageLayout(HANDLE hFile, PEImage* pOwner)
     m_Layout=LAYOUT_FLAT;
     m_pOwner=pOwner;
+
+    HANDLE hFile = pOwner->GetFileHandle();
```
The hFile argument was was always passed in as GetFileHandle(). So, we can make the change in master as well.

davidwrighton commented on src/binder/assemblybinder.cpp:
```diff
@@ -1158,13 +1166,48 @@ namespace BINDER_SPACE
+            // Bundled assemblies are treated similar to TPAs.
+            if (g_BinderVariables->fIsBundle)
+            {
+                SString candidates[] = { W(".dll"), W(".ni.dll") };
```
I believe the .ni.dll stuff is still needed for managed c++ R2R at this time.

davidwrighton commented on src/inc/corbundle.h:
```diff
@@ -0,0 +1,31 @@
+#ifndef _COR_BUNDLE_H_
+#define _COR_BUNDLE_H_
+
+class BundleInfo
```
Agreed, C++ classes are not suitable for cross binary api surface ever.

swaroop-sridhar commented on src/binder/inc/coreclrbindercommon.h:
```diff
@@ -22,7 +22,7 @@ class CLRPrivBinderCoreCLR;
 {
 public:
-    static HRESULT Init();
+    static HRESULT Init(bool isBundle);
```
Binder cannot access AppDomain members since the type is not constructed yet. The fIsBundle flag is similar to other variables such as `AppBaseURL; DynamicDirectory; DynamicBaseDirectory; AppName; AppConfigFile; AppDomainId`. But fIsBundle may not be necessary if we make BundleInfo a global.

(If we have single-file plugins in future, we may need to come back and think about this interface).

vitek-karas commented on src/binder/assemblybinder.cpp:
```diff
@@ -1158,13 +1165,48 @@ namespace BINDER_SPACE
+                for (DWORD i = 0; i < 2; i++)
+                {
+                    SString bundledName(simpleName);
+                    bundledName.Append(candidates[i]);
```
Nit: please rename the `bundleName` to probably just `assemblyFileName` or something similar - it is NOT the name of the bundle itself - confusing.

vitek-karas commented on src/binder/assemblybinder.cpp:
```diff
@@ -1158,13 +1165,48 @@ namespace BINDER_SPACE
+                    hr = GetAssembly(bundledName,
+                        TRUE,  // fIsInGAC
+                        FALSE, // fExplicitBindToNativeImage
+                        &pTPAAssembly);
```
This is sort of wrong - what we want here is to check the bundle probe - if it finds it, great - load that. But if the probe doesn't find it, we want to skip this step and instead continue below with the TPA search.
Right now I think this will actually behave differently:
* If the probe finds it - it will work fine
* If the probe doesn't find it, I think we will eventually call `CreateFile(assemblyFileName, ...)` - which I think will search for the file in current directory (didn't try this though) - we should avoid that behavior. Specifically if the assembly is still in TPA, we should use the path from TPA in this case, even if we can find the assembly also in current directory (which may not be the app's directory at all - so we might end up loading random code basically).

The short version is that this change modifies GetAssembly contract - before the path was assumed to be absolute (and it always was) - now it can be relative, but we don't limit that case to just probing the bundle.

swaroop-sridhar commented:
@vitek-karas I've addressed your feedback in the new commit.

* Fix coreclr_initialize interface
    * Pass BundleProbe as a property, instead of an additional argument to coreclr_intiialize
* Move Corbundle.h to Bundle.h
* Introduce BundleLoc to pass around Bundle location information coupled together
* Perform the Bundle-Probe close to TPA-probe (and a few other places like BindToSystem)
  so that if an assembly is not found in the bundle, it is not accidentally loaded from the disk.
* Check the bundle contents while loading assemblies through Assembly.LoadFrom(path).

vitek-karas commented on src/dlls/mscoree/unixinterface.cpp:
```diff
@@ -128,7 +130,7 @@ static void ConvertConfigPropertiesToUnicode(
     {
         propertyKeysW[propertyIndex] = StringToUnicode(propertyKeys[propertyIndex]);
-        propertyValuesW[propertyIndex] = StringToUnicode(propertyValues[propertyIndex]);
+        propertyValuesW[propertyIndex] = propertyValues[propertyIndex] != nullptr ? StringToUnicode(propertyValues[propertyIndex]) : nullptr;
```
Does this mean we try to actually convert the bundle probe pointer to UTF16 string? Isn't that like super dangerous...

vitek-karas commented on src/inc/bundle.h:
```diff
@@ -0,0 +1,62 @@
+
+class Bundle;
+
+struct BundleLoc
```
Nit: I would prefer `BundleFileLocation` or something along those lines. 'loc' to me refers to localization - and there's no need to use abbreviation here. Also "bundle location" is ambiguous - normally I would read that as "the location of the bundle".

vitek-karas commented on src/binder/inc/assembly.hpp:
```diff
@@ -26,16 +26,18 @@
+                            PEImage   **ppPEImage,
+                            PEImage   **ppNativeImage,
+                            BOOL        fExplicitBindToNativeImage,
+                            BundleLoc   bundleLoc);
```
Can we pass this as `const &` - const to make it explicit and `&` since it's two 64bit values...

swaroop-sridhar replied:
I originally used the type `const BundleLoc &`.  But this method is declared as `extern C` API, and I think the behavior of C++ reference parameters is not well-defined on such functions.  

I could use `const BundleLoc *`, but it'll need nullptr checks in a few places. Also, passing by value is convenient in a few other methods due to the use of default parameters. So, I left it passed by value here too.

vitek-karas replied:
I wonder if the `STDAPI` is still needed...
@jkotas - Two questions:
* Is there a reason the binder APIs are declared as if exported? (although they really are not)
* What is the guidance on using a bit more modern C++ - like `byref` params, `const` and so on... in CLR ?

jkotas replied:
This is left-over from fusion. Fusion was compiled as a separate thing, with its own infrastructure. Fusion did not include any VM headers, and VM did not include any fusion implementation headers. The hand shake between VM and fusion was done via `extern "C"` methods and COM interfaces.

> What is the guidance on using a bit more modern C++

As long as all compilers that we support work with it, it is ok.

swaroop-sridhar commented:
@vitek-karas I've made the corrections you suggested in https://github.com/dotnet/coreclr/pull/26504/commits/d2a0846fa8b33c2a300f05ccc9db5d154bd6527a

swaroop-sridhar commented:
@vitek-karas:  I suggest we checkin the change to single-exe branch, so that @sbomer doesn't have to work with two branches for perf-measurement. I'm happy to address more issues in subsequent checkins.

vitek-karas commented:
If it works - then sure - let's merge it...
The one thing I'm really nervous about is the usage of property value pointer as the probe pointer - not in itself, but the conversion to UTF16 which I pointed out above. That one I think we absolutely need to address, even for the prototype... (and I know that the final product will use different mechanism).

swaroop-sridhar commented:
> If it works - then sure - let's merge it...
> The one thing I'm really nervous about is the usage of property value pointer as the probe pointer - not in itself, but the conversion to UTF16 which I pointed out above. That one I think we absolutely need to address, even for the prototype... (and I know that the final product will use different mechanism).

@vitek-karas definitely agree about the Probe-pointer issue, I've already fixed that. I've also renamed `BundleLoc`. I was only talking about optimizing parameters and such -- in a later checkin. I know of a few more things that can be improved.

janvorli commented on src/dlls/mscoree/unixinterface.cpp:
```diff
@@ -127,6 +130,18 @@ static void ConvertConfigPropertiesToUnicode(
 
     for (int propertyIndex = 0; propertyIndex < propertyCount; ++propertyIndex)
     {
+        if (strcmp(propertyKeys[propertyIndex], "BUNDLE_PROBE") == 0)
```
I don't like how this makes the function do something it is not supposed to do just to save an iteration over the few properties. This function should be what its name says - just convert property keys and values to unicode. 
I would much more prefer passing the bundle probe address formatted as hex string and parse it from the string in the coreclr_initialize. This is not perf sensitive code in any way, so we should avoid hacks like passing a non-string in an array of strings.

swaroop-sridhar replied:
I certainly don't like this way of passing in the bundle-probe (or the implementation of this function). This is pretty much a hack for now -- that communicates the bundle-probe address without changing the host layer interfaces.

In the product, I think we should design (an additional) host-interface, to pass in a structure of well-typed members, instead of a list of strings. We should design the host-interface to accommodate this and other upcoming features.  Until then, my opinion is to let the hack remain, instead of polishing something that'll go away.

janvorli replied:
Ok, if this is just a temporary way, then let's keep it for now. Can you please create an issue for fixing that so that we don't forget about the need to fix it?

jkotas replied:
> we should design (an additional) host-interface

For new host-interface designs, we should also take into account preloading scenario described in https://github.com/dotnet/coreclr/pull/25725

cc @alpencolt @kbaladurin

swaroop-sridhar replied:
I created an issue to track this: https://github.com/dotnet/coreclr/issues/26752

swaroop-sridhar commented:
@janvorli I didn't notice your latest comments (was only looking at test-pass, since there wasn't much activity the last few days). I'll address your feedback in a subsequent change. Thanks.
